cache.go

```diff
@@ -0,0 +1,192 @@
+	AddFunc    func(table string, row Row)
+	UpdateFunc func(table string, old Row, new Row)
+	DeleteFunc func(table string, row Row)
+}
+
+// OnAdd calls AddFunc if it is not nil
+func (e *EventHandlerFuncs) OnAdd(table string, row Row) {
+	if e.AddFunc != nil {
+		e.AddFunc(table, row)
+	}
+}
+
+// OnUpdate calls UpdateFunc if it is not nil
+func (e *EventHandlerFuncs) OnUpdate(table string, old, new Row) {
+	if e.UpdateFunc != nil {
+		e.UpdateFunc(table, old, new)
+	}
+}
+
+// OnDelete calls DeleteFunc if it is not nil
+func (e *EventHandlerFuncs) OnDelete(table string, row Row) {
+	if e.DeleteFunc != nil {
+		e.DeleteFunc(table, row)
+	}
+}
+
+// TableCache is a collection of TableCache hashed by database name
+type TableCache struct {
+	cache         map[string]*RowCache
+	cacheMutex    sync.Mutex
+	handlers      []EventHandler
+	handlersMutex sync.Mutex
+}
+
+func newTableCache() *TableCache {
+	return &TableCache{
+		cache:         make(map[string]*RowCache),
+		cacheMutex:    sync.Mutex{},
+		handlersMutex: sync.Mutex{},
+	}
+}
+
+// Table returns the from the cache
+func (t *TableCache) Table(name string) *RowCache {
+	t.cacheMutex.Lock()
+	defer t.cacheMutex.Unlock()
+	if table, ok := t.cache[name]; ok {
+		return table
+	}
+	return nil
+}
+
+// Tables returns a list of tables
+func (t *TableCache) Tables() []string {
+	t.cacheMutex.Lock()
+	defer t.cacheMutex.Unlock()
+	var result []string
+	for k := range t.cache {
+		result = append(result, k)
+	}
+	return result
+}
+
+// Update implements the update method of the NotificationHandler interface
+// this populates the cache with new updates
+func (t *TableCache) Update(context interface{}, tableUpdates TableUpdates) {
+	if len(tableUpdates.Updates) == 0 {
+		return
+	}
+	go t.populate(tableUpdates)
+}
+
+// Locked implements the locked method of the NotificationHandler interface
+func (t *TableCache) Locked([]interface{}) {
+}
+
+// Stolen implements the stolen method of the NotificationHandler interface
+func (t *TableCache) Stolen([]interface{}) {
+}
+
+// Echo implements the echo method of the NotificationHandler interface
+func (t *TableCache) Echo([]interface{}) {
+}
+
+// Disconnected implements the disconnected method of the NotificationHandler interface
+func (t *TableCache) Disconnected() {
+}
+
+func (t *TableCache) populate(tableUpdates TableUpdates) {
+	t.cacheMutex.Lock()
+	defer t.cacheMutex.Unlock()
+	for table, updates := range tableUpdates.Updates {
+		var tCache *RowCache
+		var ok bool
+		if tCache, ok = t.cache[table]; !ok {
+			t.cache[table] = newRowCache()
+			tCache = t.cache[table]
+		}
+		tCache.mutex.Lock()
+		for uuid, row := range updates.Rows {
+			if !reflect.DeepEqual(row.New, Row{}) {
+				if existing, ok := tCache.cache[uuid]; ok {
+					if !reflect.DeepEqual(row.New, existing) {
+						tCache.cache[uuid] = row.New
+						t.handlersMutex.Lock()
+						for _, handler := range t.handlers {
+							go handler.OnUpdate(table, row.Old, row.New)
```

Comment:
Something I found useful when writing the ORM API was to add the [UUID field to all objects](https://github.com/eBay/go-ovn/blob/15eca9a4c0d78063b44b9cb844ab1c2d3347a226/model.go#L35) so callbacks would be able to access the UUID. In this case, old.Row and old.New do not have "uuid" field. Should we add it to the callback?

Reply:
I could do that... But then it would go away again when we merge #31 😉 
Certainly, the examples in this repo work fine without it... and as this is a temporary step before #31 merges I'd be happy to leave it this way for now.

Reply:
Makes sense